fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/clientlog/KeySpaceCountTree.java

```diff
@@ -83,35 +389,181 @@ protected TupleKeyCountTree newPrefixChild(@Nonnull Object prefix) {
+        int distance = 0;
+        ResolvedRecordStoreKeyspace recordStoreKeyspace = null;
+        ResolvedRecordTypeKeyspace recordTypeKeyspace = null;
+        ResolvedIndexKeyspace indexKeyspace = null;
+        for (Resolved resolved = resolvedParent; resolved != null && resolved.getDirectory() == null; resolved = resolved.getParent()) {
+            if (resolved instanceof ResolvedRecordTypeKeyspace) {
+                recordTypeKeyspace = (ResolvedRecordTypeKeyspace)resolved;
+                break;
+            }
+            if (resolved instanceof ResolvedIndexKeyspace) {
+                indexKeyspace = (ResolvedIndexKeyspace)resolved;
+                break;
+            }
+            if (resolved instanceof ResolvedRecordStoreKeyspace) {
+                recordStoreKeyspace = (ResolvedRecordStoreKeyspace)resolved;
+                break;
+            }
+            distance++;
+        }
+        if (recordStoreKeyspace != null && recordStoreKeyspace.getRecordMetaData() != null) {
+            switch (recordStoreKeyspace.getRecordStoreKeyspace()) {
+                case RECORD:
+                    if (distance == 0 && object != null && recordStoreKeyspace.getRecordMetaData().primaryKeyHasRecordTypePrefix()) {
+                        final RecordType recordType;
+                        try {
+                            recordType = recordStoreKeyspace.getRecordMetaData().getRecordTypeFromRecordTypeKey(object);
+                        } catch (RecordCoreException ex) {
+                            break;
+                        }
+                        return CompletableFuture.completedFuture(new ResolvedRecordTypeKeyspace(resolvedParent, recordType));
+                    }
+                    KeyExpression commonPrimaryKey = recordStoreKeyspace.getRecordMetaData().commonPrimaryKey();
+                    if (commonPrimaryKey != null && distance < commonPrimaryKey.getColumnSize()) {
+                        return resolvePrimaryKeyField(context, resolvedParent, object, commonPrimaryKey, distance);
+                    }
+                    break;
+                case INDEX:
+                case INDEX_SECONDARY_SPACE:
+                case INDEX_RANGE_SPACE:
+                case INDEX_UNIQUENESS_VIOLATIONS_SPACE:
+                case INDEX_BUILD_SPACE:
+                    if (distance == 0 && object != null) {
+                        final Index index;
+                        try {
+                            index = recordStoreKeyspace.getRecordMetaData().getIndexFromSubspaceKey(object);
+                        } catch (RecordCoreException ex) {
+                            break;
+                        }
+                        return CompletableFuture.completedFuture(new ResolvedIndexKeyspace(resolvedParent, index));
+                    }
+                    break;
+                default:
```

Comment:
Would it also make sense to have a `case INDEX_STATE_SPACE` that got the index from the name (instead of the subspace key)? Or do we not want to do that to protect against that changing because of how #514 could be resolved?

Reply:
Yes. Also, since the current value is the name of the index, resolving it isn't so important. It will display more or less correctly just as an unresolved value.